### PR TITLE
feat(condo): DOMA-9373 new flow to send notification about added BillingReceipts

### PR DESCRIPTION
**Pull request by VKislov**
In the main Cron task, 1 subtask is created for each `BillingIntegrationOrganizationContext`. This made it possible to abandon the use of chunk queries and also optimize the work by parallelizing data processing. Also now all queries use `find`. Additional context can be found in the task.

### CODE REVIEW

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 1]
```diff
@@ -193,20 +46,26 @@ const sendBillingReceiptsAddedNotifications = async (resendFromDt = null) => {
...
+    logger.info({ msg: 'stored date', storedDt: await redisClient.get(REDIS_LAST_DATE_KEY), taskId })
 
-    const receiptsWhere = {
-        period_in: [prevMonthStart, thisMonthStart],
-        createdAt_gt: lastDt,
+    const lastSync = new Date(lastDt).toISOString()
+    if (!lastSync) throw new Error(`Invalid last sync date: ${lastDt}`)
+    // After synchronization with integrations (RegisterBillingReceiptsService), the context is updated to put the lastReport:JSON field. 
+    // Therefore, we understand that according to the context updated after the date of the last notification mailing, there are new receipts and we fetch such contexts.
+    const BIOContexts = await find('BillingIntegrationOrganizationContext', {
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:55]: BillingContexts

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 2]
```diff
@@ -193,20 +46,26 @@ const sendBillingReceiptsAddedNotifications = async (resendFromDt = null) => {
...
+        updatedAt_gte: lastSync,
+        status: CONTEXT_FINISHED_STATUS,
         deletedAt: null,
-    }
-    await sendBillingReceiptsAddedNotificationsForPeriod(receiptsWhere, handleLastDtChange)
+    })
+
+    BIOContexts.forEach((context) => {
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:61]: same here

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 3]
```diff
@@ -193,20 +46,26 @@ const sendBillingReceiptsAddedNotifications = async (resendFromDt = null) => {
...
+        const lastReport = get(context, 'lastReport.finishTime')
+        if ( lastReport > lastSync) {
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:63]: maybe, use dayjs for compairing?

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js [hunk 4]
```diff
@@ -13,87 +14,88 @@ const {
...
-
+    let admin
+    beforeAll(async () => {
+        admin = await makeLoggedInAdminClient()
+    })
     describe('notifications', () => {
         it('sends notification of BILLING_RECEIPT_ADDED_TYPE for toPay > 0', async () => {
-            const admin = await makeLoggedInAdminClient()
-            const { receipt, resident } = await makeBillingReceiptWithResident({ toPay: '10000', toPayDetails: { charge: '1000', formula: '', balance: '9000', penalty: '0' } } )
+            const { receipt, resident, billingContext } = await makeBillingReceiptWithResident({ toPay: '10000', toPayDetails: { charge: '1000', formula: '', balance: '9000', penalty: '0' } } )
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:30]: all this test utils do not update context and do not set lastReport

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js [hunk 5]
```diff
@@ -13,87 +14,88 @@ const {
...
+
             let lastDt
             const setLastDt = (dt) => lastDt = dt
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:33]: Is there some hidden meaning to this?

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptNotifications.js [hunk 6]
```diff
@@ -11,7 +13,7 @@ const logger = getLogger('sendBillingReceiptsAddedNotifications')
 
 const sendBillingReceiptNotifications = async (context = null) => {
     const isFeatureEnabled = await featureToggleManager.isFeatureEnabled(context, SEND_BILLING_RECEIPTS_NOTIFICATIONS_TASK)
-
+    const taskId = uuid()
```
Comment (pahaz) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptNotifications.js:16]: All tasks have a `this.taskId` field.

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 7]
```diff
@@ -1,189 +1,43 @@
...
-const { sendMessage } = require('@condo/domains/notification/utils/serverSchema')
-const { ServiceConsumer } = require('@condo/domains/resident/utils/serverSchema')
+const { CONTEXT_FINISHED_STATUS } = require('@condo/domains/acquiring/constants/context')
+const { getMonthStart } = require('@condo/domains/common/utils/date')
+const { sendBillingReceiptsAddedNotificationForOrganizationContextTask } = require('@condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask')
 
 
 const REDIS_LAST_DATE_KEY = 'LAST_SEND_BILLING_RECEIPT_NOTIFICATION_CREATED_AT'
-const CHUNK_SIZE = 20
+const rLock = new Redlock([getRedisClient()])
```
Comment (pahaz) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:15]: Don't use `Redlock` it doesn't work well!

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js [hunk 8]
```diff
@@ -208,4 +210,38 @@ describe('sendBillingReceiptsAddedNotificationsForPeriod', () => {
         })
     })
 
+    describe('Real-life test cases', () => {
```
Comment (pahaz) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:213]: That's too many match changes and only one test. You should think deeply about edge cases

Comment (VKislov) [reply apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:214]: The business logic of the task has not changed. The moments of implementation have changed. Previously, the starting point was `BillingReceipt`, now `BillingContext`. Therefore, the old tests check the logic in the same way as before. The real-life test was added only to check the launch of the task after the `registerBillingReciept` mutation to check the full flow. Old tests remain relevant

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptNotifications.js [hunk 9]
```diff
@@ -21,7 +19,7 @@ const sendBillingReceiptNotifications = async (context = null) => {
     }
 
     await sendBillingReceiptsAddedNotifications()
-    await sendResidentsNoAccountNotifications()
+    // await sendResidentsNoAccountNotifications()
```
Comment (vovaaxeapolla) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptNotifications.js:22]: Needs to be uncomment

Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptNotifications.js:22]: It's better to delete this line instead of commenting

File: apps/condo/domains/common/utils/date.js [hunk 10]
```diff
@@ -10,7 +10,7 @@ const DATE_FORMAT_Z = 'YYYY-MM-DDTHH:mm:ss.SSSZZ'
  * @param shouldFormat
  * @returns {*}
  */
-const getMonthStart = (dateRaw, shouldFormat = false) => {
+const getMonthStart = (dateRaw = undefined, shouldFormat = false) => {
```
Comment (AleX83Xpert) [inline apps/condo/domains/common/utils/date.js:13]: Please add test cases for dateRaw=undefined

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 11]
```diff
@@ -1,226 +1,56 @@
...
+        updatedAt_gte: lastSync,
+        status: CONTEXT_FINISHED_STATUS,
         deletedAt: null,
+    })
+    const oldWayLastSyncDate = await redisClient.get(REDIS_LAST_DATE_KEY)
+
+    for (const context of BillingContexts) {
+        const lastReport = get(context, 'lastReport.finishTime')
+        if (!lastReport) continue
+        logger.info({ msg: 'Old way redis key', oldWayLastSyncDate, lastReport, result: dayjs(oldWayLastSyncDate).isAfter(dayjs(lastReport)) })
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:41]: Please move uncommon fields (`oldWayLastSyncDate `, `lastReport` and `result`) to `data` to prevent new indexes in logs database

File: apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js [hunk 12]
```diff
@@ -17,7 +17,7 @@ async function sendBillingReceiptNotificationsWeekends () {
 }
 
 /** PODS work according to UTC, so time in cron-tasks should also be according to UTC too. */
-const sendBillingReceiptNotificationsWorkDaysTask = createCronTask('sendBillingReceiptNotificationsWorkDaysTask', '17 7-12 * * 1-5', sendBillingReceiptNotificationsWorkDays)
+const sendBillingReceiptNotificationsWorkDaysTask = createCronTask('sendBillingReceiptNotificationsWorkDaysTask', '* * * * *', sendBillingReceiptNotificationsWorkDays)
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js:20]: `* * * * *` is not looking like "work days" :)

File: apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js [hunk 13]
```diff
@@ -17,7 +17,7 @@ async function sendBillingReceiptNotificationsWeekends () {
 }
 
 /** PODS work according to UTC, so time in cron-tasks should also be according to UTC too. */
-const sendBillingReceiptNotificationsWorkDaysTask = createCronTask('sendBillingReceiptNotificationsWorkDaysTask', '17 7-12 * * 1-5', sendBillingReceiptNotificationsWorkDays)
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js:20]: Do not forget to clear crontask within bull. As I remember, crontasks disappeared in code will not removed automatically.

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+
+/**
+ * Prepares data for sendMessage to resident on available billing receipt, then tries to send the message
+ * @param {BillingIntegrationOrganizationContext} context
+ * @param {string} lastSyncDate
+ * @returns {Promise<void>}
+ */
+async function sendBillingReceiptsAddedNotificationForOrganizationContext (context, lastSyncDate) {
+    const contextId = get(context, 'id')
+    if (!contextId) throw new Error(`Invalid BillingIntegrationOrganizationContext, cannot get context.id. Context: ${context}`)
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js:100]: You will get the '[object Object]' in error message
image

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+
+    const { keystone } = getSchemaCtx('Message')
+    const redisClient = getRedisClient()
+    const { thisMonthStart } = getStartDates()
+    const receiptCreatedAfter = lastSyncDate || thisMonthStart
+
+    const receipts = await fetchReceipts(contextId, receiptCreatedAfter)
+    logger.info({ msg: `Found ${receipts.length} receipts`, contextId })
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js:108]: It's better to set a static string as value for `msg`. And move uncommon fields to unindexed `data` field

File: apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js (hunk 12 above)
Comment (vovaaxeapolla) [inline apps/condo/domains/resident/tasks/sendBillingReceiptNotifications.task.js:20]: Return normal cron plan

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+        const addressKey = get(processedReceipt, 'account.property.addressKey')
+
+        accountsNumbers.push(accountNumber)
+        receiptAccountKeys[makeAccountKey(addressKey, accountNumber)] ||= processedReceipt
+    }
+
+    return { accountNumbers: accountsNumbers.filter(Boolean), receiptAccountKeys }
+}
+
+async function hasNewerReceipts (receipt) {
```
Comment (vovaaxeapolla) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js:154]: Is it redundant check? BillingContext.lastReport changes only if newer receipts were loaded

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+
+    const receipts = await fetchReceipts(contextId, receiptCreatedAfter)
+    logger.info({ msg: `Found ${receipts.length} receipts`, contextId })
+
+    const receiptAccountData = await prepareReceiptsData(receipts, context)
+    const serviceConsumers = await fetchServiceConsumers(context, receiptAccountData.accountNumbers)
+    const consumersByAccountKey = groupConsumersByAccountKey(serviceConsumers)
+
+    let successCount = 0
+    logger.info({ msg: 'Left receipts', receiptAccountData })
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js:115]:
```suggestion
    logger.info({ msg: 'Left receipts', data: { receiptAccountData } })
```

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+
+    for (const [key, receipt] of Object.entries(receiptAccountData.receiptAccountKeys)) {
+        const consumers = consumersByAccountKey[key]
+        if (isEmpty(consumers)) continue
+
+        successCount += await notifyConsumers(redisClient, keystone, receipt, consumers)
+    }
+
+    logger.info({ msg: 'Sent billing receipts', successCount, contextId })
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotificationForOrganizationContextTask.js:124]:
```suggestion
    logger.info({ msg: 'Sent billing receipts', data: {successCount, contextId } })
```

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js [hunk 19]
```diff
@@ -3,178 +3,226 @@
...
+
+            await registerBillingReceiptsByTestClient(admin, {
+                context: { id: billingContext1.id },
+                receipts: environment.createJSONReceipt(),
+            })
+
+            expect(resident.user.id).toEqual(resident1.user.id)
+            
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext, organization: billingContext.organization.id }, dayjs(receipt.createdAt).subtract(1, 'hour').toISOString())
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext1, organization: billingContext1.organization.id }, dayjs(receipt1.createdAt).subtract(1, 'hour').toISOString())
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:215]: What if the second task starts until the first one is finished?
Could you please add a test if such situation is possible.

Comment (vovaaxeapolla) [reply apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js:215]: Execution order is not important. I need send a push from any task but only one time

File: apps/condo/domains/common/utils/date.js (hunk 10 above)
Comment (vovaaxeapolla) [reply apps/condo/domains/common/utils/date.js:13]: Test already exists. Function is used without arguments and there are warnings about it

File: apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js [hunk 20]
```diff
@@ -1,226 +1,30 @@
...
-     */
-    const lastDt = resendFromDt ? resendFromDt.replace(' ', 'T') : await redisClient.get(REDIS_LAST_DATE_KEY) || thisMonthStart
+    const BillingContexts = await find('BillingIntegrationOrganizationContext', {
+        updatedAt_gte: lastSendDate,
+        status: CONTEXT_FINISHED_STATUS,
+        deletedAt: null,
+    })
 
-    logger.info({ msg: 'stored date', storedDt: await redisClient.get(REDIS_LAST_DATE_KEY) })
+    logger.info({ msg: 'Billing contexts for pushes', data: BillingContexts })
```
Comment (AleX83Xpert) [inline apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.js:19]: To keep variable name
```suggestion
    logger.info({ msg: 'Billing contexts for pushes', data: { BillingContexts } })
```

**Review by AleX83Xpert — APPROVED** [review]

**Review by dkoviazin — APPROVED** [review]
